components/patient-access-rules/rest/src/main/java/org/phenotips/data/permissions/rest/internal/DefaultCollaboratorResourceImpl.java

```diff
@@ -0,0 +1,187 @@
+
+        return result;
+    }
+
+    @Override
+    public Response putLevelWithJson(String json, String patientId, String collaboratorId)
+    {
+        String level;
+        try {
+            JSONObject jsonObject = JSONObject.fromObject(json);
+            level = jsonObject.getString(LEVEL);
+        } catch (Exception ex) {
+            this.logger.debug("Changing collaborator's access level failed: the JSON was not properly formatted");
+            throw new WebApplicationException(Response.Status.BAD_REQUEST);
+        }
+        return putLevel(collaboratorId.trim(), level, patientId);
+    }
+
+    @Override
+    public Response putLevelWithForm(String patientId, String collaboratorId)
+    {
+        Object levelInRequest = container.getRequest().getProperty(LEVEL);
+        if (levelInRequest instanceof String) {
+            String level = levelInRequest.toString().trim();
+            if (StringUtils.isNotBlank(level)) {
+                return putLevel(collaboratorId, level, patientId);
+            }
+        }
+        this.logger.error("The id, permissions level, or both were not provided or are invalid");
+        throw new WebApplicationException(Response.Status.BAD_REQUEST);
+    }
+
+    @Override
+    public Response deleteCollaborator(String patientId, String collaboratorId)
+    {
+        this.logger.debug(
+            "Removing collaborator with id [{}] from patient record [{}] via REST", collaboratorId, patientId);
+        // besides getting the patient, checks that the user has manage access
+        PatientAccessContext patientAccessContext = this.secureContextFactory.getContext(patientId, "manage");
+
+        PatientAccess patientAccess = patientAccessContext.getPatientAccess();
+        EntityReference collaboratorReference = this.userOrGroupResolver.resolve(collaboratorId);
+        if (collaboratorReference == null) {
+            // what would be a better status to indicate that the user/group id is not valid?
+            // ideally, the status page should show some sort of a message indicating that the id was not found
+            throw new WebApplicationException(Response.Status.NOT_FOUND);
+        }
+
+        if (!patientAccess.removeCollaborator(collaboratorReference)) {
+            this.logger.error("Could not remove collaborator [{}] from patient record [{}]", collaboratorId, patientId);
+            throw new WebApplicationException(Response.Status.INTERNAL_SERVER_ERROR);
+        }
+
+        return Response.ok().build();
+    }
+
+    private CollaboratorRepresentation createCollaboratorRepresentation(Patient patient, String id,
+        PatientAccess patientAccess)
+    {
+        String collaboratorId = id.trim();
+        // check if the space reference is used more than once in this class
+        EntityReference collaboratorReference = this.userOrGroupResolver.resolve(collaboratorId);
+        if (collaboratorReference == null) {
```

Comment:
Maybe it's better to check if `collaboratorReference != null`, so there would be fewer exit points from the method.